exec/lifecycle/src/runLifecycleHook.ts

```diff
@@ -157,6 +162,14 @@ Please unset the script-shell option, or configure it to a .exe instead.
   }
 }
 
+export async function callModifyBinPaths (
+  opts: Pick<RunLifecycleHookOptions, 'extraBinPaths' | 'modifyBinPaths'>,
+  manifest: ProjectManifest
+): Promise<string[]> {
+  const modifyBinPaths: ModifyBinPaths = opts.modifyBinPaths ?? ((_, extraBinPaths) => extraBinPaths)
+  return modifyBinPaths(manifest, opts.extraBinPaths ?? [])
+}
+
```

Comment:
this function looks confusing. Maybe revert these hacks and just add some in-memory caching to the prepareExecutionEnv function instead, so that it doesn't start to download Node.js multiple times. Instead, wait for the already started download to finish.

Reply:
I have added a cache object to `createPrepareExecutionEnv` (formerly called `createBinPathsModifierBasedOnManifest`) in `node.ts`: https://github.com/pnpm/pnpm/pull/8277/commits/11c944d95635f624b0a130e9343124de2e86412b